File: Assets/_Project/Scripts/Core/Events/GameEvents.cs
using UnityEngine;

namespace VampireSurvivors.Core
{
    /// <summary>
    /// 敵人被擊殺事件 — 當敵人生命值歸零時發布。
    /// </summary>
    public struct EnemyKilledEvent : IGameEvent
    {
        /// <summary>敵人死亡位置。</summary>
        public Vector2 Position;

        /// <summary>被擊殺敵人的設定檔。</summary>
        public EnemyConfigSO Config;
    }

    /// <summary>
    /// 玩家升級事件 — 當玩家經驗值達到升級門檻時發布。
    /// </summary>
    public struct PlayerLevelUpEvent : IGameEvent
    {
        /// <summary>升級後的新等級。</summary>
        public int NewLevel;
    }

    /// <summary>
    /// 玩家受傷事件 — 當玩家角色受到傷害時發布。
    /// </summary>
    public struct PlayerDamagedEvent : IGameEvent
    {
        /// <summary>本次受到的傷害量。</summary>
        public float Damage;

        /// <summary>受傷後剩餘的生命值。</summary>
        public float RemainingHP;
    }

    /// <summary>
    /// 遊戲結束事件 — 當玩家角色死亡或遊戲結束時發布。
    /// </summary>
    public struct GameOverEvent : IGameEvent
    {
        /// <summary>本局存活時間（秒）。</summary>
        public float SurvivalTime;

        /// <summary>本局擊殺敵人總數。</summary>
        public int KillCount;

        /// <summary>本局獲得的金幣數量。</summary>
        public int Gold;
    }

    /// <summary>
    /// 武器進化事件 — 當武器完成進化時發布。
    /// </summary>
    public struct WeaponEvolvedEvent : IGameEvent
    {
        /// <summary>進化武器的識別碼。</summary>
        public string WeaponId;
    }

    /// <summary>
    /// 經驗值收集事件 — 當玩家拾取經驗寶石時發布。
    /// </summary>
    public struct XPCollectedEvent : IGameEvent
    {
        /// <summary>收集的經驗值數量。</summary>
        public float Amount;
    }

    /// <summary>
    /// 遊戲開始事件 — 當新一局遊戲開始時發布。
    /// </summary>
    public struct GameStartEvent : IGameEvent
    {
        /// <summary>玩家選擇的角色設定檔。</summary>
        public CharacterConfigSO Character;

        /// <summary>玩家選擇的地圖索引。</summary>
        public int MapIndex;
    }

    /// <summary>
    /// 武器攻擊事件 — 當武器發動攻擊時發布。
    /// </summary>
    public struct WeaponAttackEvent : IGameEvent
    {
        /// <summary>發動攻擊的武器識別碼。</summary>
        public string WeaponId;
    }

    /// <summary>
    /// 敵人受傷事件 — 當敵人受到傷害時發布。
    /// </summary>
    public struct EnemyDamagedEvent : IGameEvent
    {
        /// <summary>受傷敵人的位置。</summary>
        public Vector2 Position;

        /// <summary>本次受到的傷害量。</summary>
        public float Damage;
    }

    /// <summary>
    /// 勝利事件 — 當玩家擊敗最終 Boss 時發布。
    /// </summary>
    public struct VictoryEvent : IGameEvent
    {
        /// <summary>本局存活時間（秒）。</summary>
        public float SurvivalTime;

        /// <summary>本局擊殺敵人總數。</summary>
        public int KillCount;
    }
}

File: src/core/events/GameEvents.ts
// === core/events/GameEvents.ts ===

export interface EnemyKilledEvent {
  position: { x: number; y: number };
  enemyId: string;
  xpValue: number;
}

export interface PlayerLevelUpEvent {
  newLevel: number;
}

export interface PlayerDamagedEvent {
  damage: number;
  remainingHP: number;
}

export interface GameOverEvent {
  survivalTime: number;
  killCount: number;
  gold: number;
  maxLevel: number;
}

export interface GameVictoryEvent {
  survivalTime: number;
  killCount: number;
  gold: number;
}

export interface XPCollectedEvent {
  amount: number;
}

export interface WeaponEvolvedEvent {
  weaponId: string;
  evolvedWeaponId: string;
}

export interface WaveChangedEvent {
  waveIndex: number;
  difficultyMultiplier: number;
}

export interface EnemyDamagedEvent {
  position: { x: number; y: number };
  damage: number;
  enemyId: string;
}

// 事件名稱常數
export const GameEventNames = {
  ENEMY_KILLED: 'enemy:killed',
  ENEMY_DAMAGED: 'enemy:damaged',
  PLAYER_LEVEL_UP: 'player:levelUp',
  PLAYER_DAMAGED: 'player:damaged',
  GAME_OVER: 'game:over',
  GAME_VICTORY: 'game:victory',
  XP_COLLECTED: 'xp:collected',
  WEAPON_EVOLVED: 'weapon:evolved',
  WAVE_CHANGED: 'wave:changed',
  GAME_PAUSED: 'game:paused',
  GAME_RESUMED: 'game:resumed',
} as const;
